pytorch_lightning/trainer/connectors/logger_connector/result.py

```diff
@@ -279,9 +282,16 @@ def extra(self, extra: Mapping[str, Any]) -> None:
 
         def check_fn(v):
             if v.grad_fn is not None:
-                raise MisconfigurationException(f'You returned a tensor with `grad_fn`. The extra values are {extra}')
+                # raise MisconfigurationException(f'You returned a tensor with `grad_fn`. The extra values are {extra}')
```

Comment:
Do we need to update the `metrics(...)` function to also detach when returning the collection?

```
            # populate logging metrics
            if result_metric.meta.logger:
                metrics[MetricSource.LOG][forked_name] = value.detach()

            # populate callback metrics. callback metrics don't take `_step` forked metrics
            if self.training or result_metric.meta.on_epoch and not on_step:
                metrics[MetricSource.CALLBACK][name] = value.detach()
                metrics[MetricSource.CALLBACK][forked_name] = value.detach()
```

I'm not super familiar with the old code, so I'm not sure if automatically detaching was something we did only for `extra` or for everything else as well? I suspect it must have been for everything since we have some tests internally that `deepcopy` the LMs and this does not work when not detaching the tensors.

Reply:
I think we were removing the _results parameter from the pl_module before deepcopying which doesn't exist anymore. Might just need to see what its saved now as

Reply:
Yep, updated the corresponding code internally. That addresses our quantization piece, but I think this change is still needed for other libraries. There's an internal test that fails because we do a deepcopy in model_checkpoint.py:

```
# from model_checkpoint.py

def _monitor_candidates(self, trainer: 'pl.Trainer', epoch: int, step: int) -> Dict[str, _METRIC]:
        monitor_candidates = deepcopy(trainer.logger_connector.callback_metrics)
        monitor_candidates.update(epoch=epoch, step=step)
        return monitor_candidates
```

The callback metrics dictionary there includes `train_loss` and `train_loss_step` which both have `require_grad=True` at that point (so deepcopy is not possible).

The LM seems to do the right thing though, since all it does is return a single loss tensor in `training_step`.

Reply:
Are you using `self.log(..., enable_graph=True)`?

Reply:
Ah, you're right! Yes, the LM is using this option :) Will try removing it now.

Reply:
Great. Merging this. If there's any further issue with `enable_graph`, we can deal with it separately